AudioPlayer/src/main/kotlin/com.github.diamondminer88.plugins/AudioPlayer.kt

```diff
@@ -34,241 +34,363 @@ import java.util.*
+    private val attachmentCardId = Utils.getResId("chat_list_item_attachment_card", "id")
+    private val validFileExtensions = arrayOf(
+        "webm", "mp3", "aac", "m4a", "wav", "flac", "wma", "opus", "ogg"
+    )
+
+    private val allPlayerBarResets = mutableListOf<() -> Unit>()
+
+    private var globalCurrentPlayer: MediaPlayer? = null
+    private var globalCleanup: (() -> Unit)? = null
+
+    private fun isAudioFile(filename: String?): Boolean {
+        if (filename == null) return false
+        val ext = filename.substringAfterLast('.', "").lowercase(Locale.ROOT)
+        return validFileExtensions.contains(ext)
+    }
+
+    private fun msToTime(ms: Long): String {
+        val hrs = ms / 3_600_000
+        val mins = ms / 60000
+        val secs = ms / 1000 % 60
+
+        return if (hrs == 0L)
+            String.format("%d:%02d", mins, secs)
+        else
+            String.format("%d:%d:%02d", hrs, mins, secs)
+    }
+
+    private fun stopCurrentPlayer() {
+        try { globalCleanup?.invoke() } catch (_: Exception) {}
+        try { globalCurrentPlayer?.stop() } catch (_: Exception) {}
+        try { globalCurrentPlayer?.release() } catch (_: Exception) {}
+        globalCurrentPlayer = null
+        globalCleanup = null
+
+        allPlayerBarResets.forEach { it() }
+    }
+
+    override fun start(context: Context) {
+        patcher.after<WidgetChatListAdapterItemAttachment>(
+            "configureFileData",
+            MessageAttachment::class.java,
+            MessageRenderContext::class.java
+        ) {
+            val messageAttachment = it.args[0] as MessageAttachment
+            val root = WidgetChatListAdapterItemAttachment.`access$getBinding$p`(this).root as ConstraintLayout
+            val card = root.findViewById<MaterialCardView>(attachmentCardId)
+            val ctx = root.context
+
+            card.findViewById<MaterialCardView>(playerBarId)?.let { card.removeView(it) }
+            val loadingBarId = playerBarId + 1
+            card.findViewById<ProgressBar>(loadingBarId)?.let { card.removeView(it) }
+
+            if (!isAudioFile(messageAttachment.filename)) return@after
+
+            val loadingBar = ProgressBar(ctx, null, android.R.attr.progressBarStyleHorizontal).apply {
+                id = loadingBarId
+                isIndeterminate = true
+                layoutParams = FrameLayout.LayoutParams(MATCH_PARENT, 6.dp).apply {
+                    gravity = Gravity.BOTTOM
+                }
+            }
+            card.addView(loadingBar)
+
+            Utils.threadPool.execute {
+                val isOgg = messageAttachment.filename.lowercase(Locale.ROOT).endsWith(".ogg")
+                val localOggFile = if (isOgg) File(ctx.cacheDir, "audio.ogg") else null
+                if (isOgg) {
+                    localOggFile!!.deleteOnExit()
+                    Http.simpleDownload(messageAttachment.url, localOggFile)
+                }
```

Comment:
Files should not be downloaded automatically without user interactions, and especially not like this.
IIRC `configureFileData` runs every single time the view is bound, meaning it would be re-downloaded every single time it scrolls into view. Not sure on this though.

This also creates a race condition since multiple attachments would be downloaded at the same time to the same file.